src/storage/fs_storage.rs

```diff
@@ -0,0 +1,457 @@
+    write_compacted_term(&data_dir, 0)?;
+    write_first_index(&data_dir, 1)?;
+    write_last_index(&data_dir, 0)?;
+    write_applied(&data_dir, 0)?;
+    write_hard_state(&data_dir, &HardState::new())?;
+    write_snapshot(&data_dir, &Snapshot::new())
+}
+
+fn write_hard_state<P: AsRef<Path>>(data_dir: P, hard_state: &HardState) -> io::Result<()> {
+    write_pb_to_file(data_dir.as_ref().join("hard_state"), hard_state)
+}
+
+fn write_snapshot<P: AsRef<Path>>(data_dir: P, snapshot: &Snapshot) -> io::Result<()> {
+    write_pb_to_file(data_dir.as_ref().join("snapshot"), snapshot)
+}
+
+fn write_compacted_term<P: AsRef<Path>>(data_dir: P, term: u64) -> io::Result<()> {
+    write_u64_to_file(data_dir.as_ref().join("term"), term)
+}
+
+fn write_first_index<P: AsRef<Path>>(data_dir: P, first: u64) -> io::Result<()> {
+    write_u64_to_file(data_dir.as_ref().join("first"), first)
+}
+
+fn write_last_index<P: AsRef<Path>>(data_dir: P, last: u64) -> io::Result<()> {
+    write_u64_to_file(data_dir.as_ref().join("last"), last)
+}
+
+fn write_applied<P: AsRef<Path>>(data_dir: P, applied: u64) -> io::Result<()> {
+    write_u64_to_file(data_dir.as_ref().join("applied"), applied)
+}
+
+fn write_entry<P: AsRef<Path>>(entries_dir: P, entry: &Entry) -> io::Result<()> {
+    write_pb_to_file(entries_dir.as_ref().join(format!("{}", entry.index)), entry)
+}
+
+fn remove_entry<P: AsRef<Path>>(entries_dir: P, index: u64) -> io::Result<()> {
+    fs::remove_file(entries_dir.as_ref().join(format!("{}", index)))
+}
+
+fn read_pb_from_file<P: AsRef<Path>, O: ProtobufMessage>(path: P) -> io::Result<O> {
+    fs::read(path).and_then(|payload| {
+        protobuf::parse_from_bytes(&payload)
+            .map_err(|err| io::Error::new(io::ErrorKind::InvalidData, err))
+    })
+}
+
+fn write_pb_to_file<P: AsRef<Path>, O: ProtobufMessage>(path: P, pb: &O) -> io::Result<()> {
+    pb.write_to_bytes()
+        .map_err(|err| io::Error::new(io::ErrorKind::Other, err))
+        .and_then(|payload| fs::write(path, &payload))
+}
+
+fn read_and_map_dir<T, P: AsRef<Path>>(
+    path: P,
+    f: fn(fs::DirEntry) -> io::Result<T>,
+) -> io::Result<Vec<T>> {
+    Ok(fs::read_dir(path)?
+        .map(|result| result.map(f)?)
+        .collect::<Result<Vec<T>, io::Error>>()?)
+}
+
+fn read_u64_from_file<P: AsRef<Path>>(path: P) -> io::Result<u64> {
+    const SIZE: usize = mem::size_of::<u64>();
+    let payload = fs::read(path)?;
+    if payload.len() != SIZE {
+        return Err(io::Error::new(
+            io::ErrorKind::InvalidData,
+            "File does not contain u64",
+        ));
+    }
+    let mut buf: [u8; SIZE] = [0; SIZE];
+    (&mut buf[..]).copy_from_slice(&payload);
+    Ok(u64_from_bytes(buf))
+}
+
+fn write_u64_to_file<P: AsRef<Path>>(path: P, term: u64) -> io::Result<()> {
+    fs::write(path, &u64_to_bytes(term))
+}
+
+#[inline]
+fn u64_to_bytes(this: u64) -> [u8; mem::size_of::<u64>()] {
+    unsafe { mem::transmute(this) }
```

Comment:
To be completely sure you write and read little-endian 64-bit words, you might need [`byteorder`](https://docs.rs/byteorder/1.3.2/byteorder/).

Reply:
Well if you're going that far, then you need to think about the endianness of the filesystem also?